sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala

```diff
@@ -381,4 +388,187 @@ class StreamingAggregationSuite extends StateStoreMetricsTest
+          CheckLastBatch(3),
+          StopStream
+        )
+      }
+    }
+  }
+
+  test("SPARK-21977: coalesce(1) should still be repartitioned when it has keyExpressions") {
+    val inputSource = new BlockRDDBackedSource(spark)
+    MockSourceProvider.withMockSources(inputSource) {
+      withTempDir { tempDir =>
+
+        def createDf(partitions: Int): Dataset[(Long, Long)] = {
+          spark.readStream
+            .format((new MockSourceProvider).getClass.getCanonicalName)
+            .load()
+            .coalesce(partitions)
+            .groupBy('a % 1) // just to give it a fake key
+            .count()
+            .as[(Long, Long)]
+        }
+
+        testStream(createDf(1), Complete())(
+          StartStream(checkpointLocation = tempDir.getAbsolutePath),
+          AddBlockData(inputSource, Seq(1)),
+          CheckLastBatch((0L, 1L)),
+          AssertOnQuery("Verify addition of exchange operator") { se =>
+            checkAggregationChain(
+              se,
+              expectShuffling = true,
+              spark.sessionState.conf.numShufflePartitions)
+          },
+          StopStream
+        )
+
+        testStream(createDf(2), Complete())(
+          StartStream(checkpointLocation = tempDir.getAbsolutePath),
+          Execute(se => se.processAllAvailable()),
+          AddBlockData(inputSource, Seq(2), Seq(3), Seq(4)),
+          CheckLastBatch((0L, 4L)),
+          AssertOnQuery("Verify no exchange added") { se =>
+            checkAggregationChain(
+              se,
+              expectShuffling = false,
+              spark.sessionState.conf.numShufflePartitions)
+          },
+          AddBlockData(inputSource),
+          CheckLastBatch((0L, 4L)),
+          StopStream
+        )
+      }
+    }
+  }
+}
+
+/**
+ * A Streaming Source that is backed by a BlockRDD and that can create RDDs with 0 blocks at will.
+ */
+class BlockRDDBackedSource(spark: SparkSession) extends Source {
+  var counter = 0L
+  private val blockMgr = SparkEnv.get.blockManager
+  private var blocks: Seq[BlockId] = Seq.empty
+
+  private var streamLock: CountDownLatch = new CountDownLatch(1)
+
+  def addData(data: Int*): Unit = {
+    if (streamLock.getCount == 0) {
+      streamLock = new CountDownLatch(1)
+    }
+    synchronized {
+      if (data.nonEmpty) {
+        val id = TestBlockId(counter.toString)
+        blockMgr.putIterator(id, data.iterator, StorageLevel.MEMORY_ONLY)
+        blocks ++= id :: Nil
+      }
+      counter += 1
+    }
+  }
+
+  def releaseLock(): Unit = streamLock.countDown()
+
+  override def getOffset: Option[Offset] = {
+    streamLock.await()
+    synchronized {
+      if (counter == 0) None else Some(LongOffset(counter))
+    }
+  }
+
+  override def getBatch(start: Option[Offset], end: Offset): DataFrame = synchronized {
+    val rdd = new BlockRDD[Int](spark.sparkContext, blocks.toArray)
+      .map(i => InternalRow(i)) // we don't really care about the values in this test
```

Comment:
we do care about the values in the test that has `.groupBy('a % 1)`

Reply:
not really since the grouping key is consistently 1